[Tan Li Xin] iP

tlx02 opened the pull request:
### **Byu**
> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes/))

Byu is a friendly and efficient chatbot that helps to keep track of your tasks.

It's,
- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,
1. download it from [here](https://github.com/nus-cs2103-AY2223S1/ip/pull/338)
2. double-click it
3. add your tasks
4. let it manage your tasks for you :star_struck:

Features:

- [x] Managing tasks
- [ ] Reminders (coming soon)
## 

If you are a Java programmer, you can use it to practice Java too! Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Commit:
Added 'Greet, Echo, Exit' features

Commit:
Added 'Add' and 'List' features

Commit:
Added ability to mark tasks as done

Commit:
Added support for tracking three types of tasks: ToDos, Events, Deadlines

Commit:
text-ui-test

Commit:
Added handling of errors

Commit:
Added delete function

Commit:
Used Enums

Commit:
Add save function

Commit:
Add understanding of dates and times for deadlines

Commit:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Add save function

Commit:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Add understanding of dates and times for deadlines

# Conflicts:
#	src/main/java/Deadlines.java
#	src/main/java/Duke.java
#	src/main/java/Events.java

Commit:
Use more OOP

Commit:
Divide classes into packages

Commit:
Merge branch 'cs2103add-gradle-support' into add-gradle-support

* cs2103add-gradle-support:
  Add Gradle support

Commit:
Merge branch 'add-gradle-support'

* add-gradle-support:
  Add Gradle support

Commit:
Use Gradle

Commit:
Add JUnit tests

Commit:
Create JAR file

Commit:
Add JavaDoc comments

Commit:
Follow the coding standard

Commit:
Give users a way to find a task by searching for a keyword

Commit:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add JavaDoc comments

Commit:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Follow the coding standard

# Conflicts:
#	Duke.txt
#	src/main/java/byu/Byu.java
#	src/main/java/byu/FileReader.java
#	src/main/java/byu/ToDoList.java
#	src/main/java/byu/Ui.java
#	src/main/java/commands/AddCommand.java
#	src/main/java/task/Task.java

Commit:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Give users a way to find a task by searching for a keyword

# Conflicts:
#	src/main/java/byu/Parser.java

KeithPJX commented on Duke.txt:
```diff
@@ -0,0 +1,3 @@
+T | 1 | sleep
+T | 0 | jump
+T | 0 | jog
```
I think this file is not supposed to be in the repository. It is in the .gitignore.

KeithPJX commented on src/main/java/byu/Byu.java:
```diff
@@ -0,0 +1,54 @@
+package byu;
+
+import commands.Command;
+import exceptions.DukeException;
+import java.io.IOException;
```
Is the ordering of the import statements correct? Might need a space. [link to import section of coding standard](https://se-education.org/guides/conventions/java/intermediate.html#package-and-import-statements)

KeithPJX commented on src/main/java/byu/FileReader.java:
```diff
@@ -0,0 +1,73 @@
+
+import java.util.Scanner;
+
+import task.*;
```
IDE might automatically do a wildcard import if you import everything from that package. Not sure if this will be allowed in the given coding standard.

KeithPJX commented on src/main/java/byu/Parser.java:
```diff
@@ -0,0 +1,147 @@
+package byu;
+
+import commands.*;
```
Another wildcard import as mentioned earlier. Not sure if this is allowed.

KeithPJX reviewed (commented):
Looks good. However, found some coding standard violations.

KeithPJX commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,54 @@
+public class Task {
+
+    private String name;
+    private boolean done;
```
I think this does not meet the coding standard. Maybe isDone will be better?

Commit:
Fix coding standard

Commit:
Fix bugs and javadocs

Commit:
Use checkStyle to fix coding style violations

Commit:
Add javadoc for DukeException

Commit:
Merge branch 'branch-A-CheckStyle'

* branch-A-CheckStyle:
  Add javadoc for DukeException

Commit:
Add GUI to Byu and update Byu.java to return String response from user input

Commit:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Add GUI to Byu and update Byu.java to return String response from user input

tanxuanyun commented on src/main/java/byu/Byu.java:
```diff
@@ -0,0 +1,90 @@
+        while (!isExit) {
+            try {
+                String fullCommand = ui.readCommand();
+                Command c = Parser.parse(fullCommand);
```
Perhaps c can be named more meaningfully since it is use quite a lot?

tanxuanyun commented on src/main/java/byu/Parser.java:
```diff
@@ -0,0 +1,206 @@
+            c = prepareDelete(description);
+            break;
+        case BYE:
+            c = new ByeCommand();
+            break;
+        case LIST:
+            c = new ListCommand();
+            break;
+        case FIND:
+            c = new FindCommand(description);
+            break;
+        default:
+            c = new NextCommand();
```
I like how the abstraction is done here

tanxuanyun commented on src/main/java/byu/Ui.java:
```diff
@@ -0,0 +1,55 @@
+    /**
+     * Prints the welcome message.
+     */
+    public void showWelcome() {
```
I like how the function are named accurately and at a sufficient level of detail.

tanxuanyun commented on src/main/java/byu/Byu.java:
```diff
@@ -0,0 +1,90 @@
+    /**
+     * Creates an instance of Byu with data from the file.
+     */
+    public Byu() {
```
I like how the code is structured logically

Commit:
Update GUI properties

Commit:
Use assertions in list() and showWelcome() methods.

There are some assumptions in the code. These assumptions are not specified.

Using assertions to document assumptions identifies possible bugs during runtime.

Let's use assertions to document important assumptions that should hold at various points in the code.

Commit:
Improve code quality

Some classes have long methods and complicated expressions, decreasing readability.

Refactoring helps to shorten methods and simplify expressions.

Let's extract code fragments that can be grouped together and consolidate conditional expressions.

Commit:
Use the Streams feature in list() method of TaskList.java

tlx02 committed:
Merge pull request #1 from tlx02/branch-A-Assertions

Add assertions to list() and showWelcome() methods

Commit:
Merge branch 'master' of https://github.com/tlx02/ip

* 'master' of https://github.com/tlx02/ip:
  Use assertions in list() and showWelcome() methods.

Commit:
Merge branch 'master' into branch-A-CodeQuality

* master:
  Use assertions in list() and showWelcome() methods.

# Conflicts:
#	src/main/java/byu/Ui.java

Commit:
Merge branch 'master' into branch-A-Streams

* master:
  Use assertions in list() and showWelcome() methods.

tlx02 committed:
Merge pull request #2 from tlx02/branch-A-CodeQuality

Improve code quality

Commit:
Merge branch 'master' of https://github.com/tlx02/ip

* 'master' of https://github.com/tlx02/ip:
  Improve code quality

Commit:
Merge branch 'master' into branch-A-Streams

* master:
  Improve code quality

tlx02 committed:
Merge pull request #3 from tlx02/branch-A-Streams

Use the Streams feature in list() method of TaskList.java

Commit:
Merge branch 'master' of https://github.com/tlx02/ip

* 'master' of https://github.com/tlx02/ip:
  Use the Streams feature in list() method of TaskList.java

Commit:
Fix bugs in saving of tasks to Duke.txt

Currently, changes are only saved when user inputs the bye command.

By saving the changes to Duke.txt after each command, the changes will
not be lost even if the user exits the program without typing 'bye'.

Let's update Duke.txt after each command given by the user.

Commit:
Add extension to detect duplicates

Currently, same tasks can be added, resulting in meaningless duplicates.

By adding this extension, users will be informed when a task already
exists and the same task will not be added more than once.

Let's detect duplicates and print a message to the user when the same
task is added.

Commit:
Merge branch 'branch-C-DetectDuplicates'

* branch-C-DetectDuplicates:
  Add extension to detect duplicates

Commit:
Improve code quality for files in the 'task' package
and check for valid deadline in Parser.java instead of Deadline.java

Commit:
Improve code quality and coding standard for files in the 'exception'
package

Commit:
Improve code quality and coding standard for files in the 'commands'
package

Commit:
Implement the generation of response in each command class instead of
TaskList.java

Commit:
Improve code quality and coding standard for files in the 'byu' package

Commit:
Improve GUI

Currently, the GUI has little padding for the labels and the square
images are not visually pleasing.

By improving the GUI, it improves the user experience.

Let's:
* make the corners of the labels round
* increase paddings for labels
* change shape of images to circle

Commit:
Give Byu a unique personality by changing its responses to the commands

Commit:
Store dates and times of events as LocalDateTime objects

Commit:
Add and edit JUnit tests for Parser.java, TaskList.java and
Deadline.java

Commit:
Repackage files and improve code quality

tlx02 committed:
Update README.md

tlx02 committed:
Add a screenshot of the product

Commit:
Fix bugs in JUnit tests

tlx02 committed:
Delete Byu.png

tlx02 committed:
Add files via upload

Commit:
Update Byu to exit when user inputs 'bye'

Commit:
Merge branch 'master' of https://github.com/tlx02/ip

* 'master' of https://github.com/tlx02/ip:
  Add files via upload
  Delete Byu.png
  Add a screenshot of the product
  Update README.md

tlx02 committed:
Update README.md

tlx02 committed:
Update README.md

Commit:
Merge branch 'master' of https://github.com/tlx02/ip

* 'master' of https://github.com/tlx02/ip:
  Update README.md
  Update README.md

Commit:
Remove text-ui-test

tlx02 committed:
Update README.md

Commit:
Merge branch 'master' of https://github.com/tlx02/ip

* 'master' of https://github.com/tlx02/ip:
  Update README.md

Commit:
Change images

Commit:
Fix window resizing issues

Commit:
Add title for window

Commit:
Add help command

Commit:
Refactor code and improve code quality for the commands

Commit:
Refactor and improve code quality for remaining classes.

Commit:
Remove HelpWindow.java class

Commit:
Add image for help window
